File: jas/panels/character_panel_state_test.py
"""Tests for the Character panel apply-to-selection pipeline."""

from __future__ import annotations

import pytest

from panels.character_panel_state import (
    _attrs_from_panel,
    _fmt_num,
    apply_character_panel_to_selection,
    subscribe,
)


# ── _fmt_num ─────────────────────────────────────────────────

class TestFmtNum:
    def test_integer_no_decimal(self):
        assert _fmt_num(12.0) == "12"

    def test_fraction_trimmed(self):
        assert _fmt_num(14.4) == "14.4"

    def test_trailing_zeros_trimmed(self):
        assert _fmt_num(14.5000) == "14.5"

    def test_negative(self):
        assert _fmt_num(-3.0) == "-3"


# ── _attrs_from_panel — text_decoration ──────────────────────

class TestTextDecoration:
    def test_neither(self):
        attrs = _attrs_from_panel({"underline": False, "strikethrough": False})
        assert attrs["text_decoration"] == ""

    def test_underline_only(self):
        attrs = _attrs_from_panel({"underline": True})
        assert attrs["text_decoration"] == "underline"

    def test_strikethrough_only(self):
        attrs = _attrs_from_panel({"strikethrough": True})
        assert attrs["text_decoration"] == "line-through"

    def test_both_alphabetical(self):
        attrs = _attrs_from_panel({"underline": True, "strikethrough": True})
        assert attrs["text_decoration"] == "line-through underline"


# ── _attrs_from_panel — caps / sub / super ───────────────────

class TestCapsAndBaseline:
    def test_all_caps_wins_over_small_caps(self):
        attrs = _attrs_from_panel({"all_caps": True, "small_caps": True})
        assert attrs["text_transform"] == "uppercase"
        assert attrs["font_variant"] == ""

    def test_small_caps_when_all_caps_off(self):
        attrs = _attrs_from_panel({"small_caps": True})
        assert attrs["text_transform"] == ""
        assert attrs["font_variant"] == "small-caps"

    def test_super_wins_over_numeric(self):
        attrs = _attrs_from_panel({"superscript": True, "baseline_shift": 5})
        assert attrs["baseline_shift"] == "super"

    def test_sub_when_super_off(self):
        attrs = _attrs_from_panel({"subscript": True})
        assert attrs["baseline_shift"] == "sub"

    def test_numeric_baseline_shift(self):
        attrs = _attrs_from_panel({"baseline_shift": 3})
        assert attrs["baseline_shift"] == "3pt"

    def test_zero_baseline_shift_empty(self):
        attrs = _attrs_from_panel({"baseline_shift": 0})
        assert attrs["baseline_shift"] == ""


# ── _attrs_from_panel — style_name parsing ───────────────────

class TestStyleName:
    def test_regular(self):
        attrs = _attrs_from_panel({"style_name": "Regular"})
        assert attrs["font_weight"] == "normal"
        assert attrs["font_style"] == "normal"

    def test_italic(self):
        attrs = _attrs_from_panel({"style_name": "Italic"})
        assert attrs["font_weight"] == "normal"
        assert attrs["font_style"] == "italic"

    def test_bold(self):
        attrs = _attrs_from_panel({"style_name": "Bold"})
        assert attrs["font_weight"] == "bold"
        assert attrs["font_style"] == "normal"

    def test_bold_italic(self):
        attrs = _attrs_from_panel({"style_name": "Bold Italic"})
        assert attrs["font_weight"] == "bold"
        assert attrs["font_style"] == "italic"

    def test_unknown_style_no_font_keys(self):
        """Unknown style names leave weight/style untouched."""
        attrs = _attrs_from_panel({"style_name": "Something Weird"})
        assert "font_weight" not in attrs
        assert "font_style" not in attrs


# ── _attrs_from_panel — leading / tracking ───────────────────

class TestLeadingTracking:
    def test_leading_at_auto_default_empties(self):
        attrs = _attrs_from_panel({"font_size": 12, "leading": 14.4})
        assert attrs["line_height"] == ""

    def test_leading_off_auto(self):
        attrs = _attrs_from_panel({"font_size": 12, "leading": 20})
        assert attrs["line_height"] == "20pt"

    def test_tracking_zero_empties(self):
        attrs = _attrs_from_panel({"tracking": 0})
        assert attrs["letter_spacing"] == ""

    def test_tracking_positive(self):
        attrs = _attrs_from_panel({"tracking": 25})  # 25/1000 em
        assert attrs["letter_spacing"] == "0.025em"

    def test_kerning_zero_empties(self):
        attrs = _attrs_from_panel({"kerning": 0})
        assert attrs["kerning"] == ""

    def test_kerning_positive(self):
        attrs = _attrs_from_panel({"kerning": 50})
        assert attrs["kerning"] == "0.05em"

    def test_kerning_numeric_string(self):
        # combo_box commits values as strings — numeric form still
        # converts to "{N}em".
        attrs = _attrs_from_panel({"kerning": "25"})
        assert attrs["kerning"] == "0.025em"

    def test_kerning_named_modes_pass_through(self):
        assert _attrs_from_panel({"kerning": "Optical"})["kerning"] == "Optical"
        assert _attrs_from_panel({"kerning": "Metrics"})["kerning"] == "Metrics"

    def test_kerning_auto_empties(self):
        # Auto / "" / "0" all round-trip to an empty element attribute.
        assert _attrs_from_panel({"kerning": "Auto"})["kerning"] == ""
        assert _attrs_from_panel({"kerning": "0"})["kerning"] == ""
        assert _attrs_from_panel({"kerning": ""})["kerning"] == ""


# ── _attrs_from_panel — rotation / scale ─────────────────────

class TestRotationScale:
    def test_rotation_zero_empties(self):
        attrs = _attrs_from_panel({"character_rotation": 0})
        assert attrs["rotate"] == ""

    def test_rotation_numeric(self):
        attrs = _attrs_from_panel({"character_rotation": 15})
        assert attrs["rotate"] == "15"

    def test_scale_identity_empties(self):
        attrs = _attrs_from_panel({"horizontal_scale": 100, "vertical_scale": 100})
        assert attrs["horizontal_scale"] == ""
        assert attrs["vertical_scale"] == ""

    def test_scale_off_identity(self):
        attrs = _attrs_from_panel({"horizontal_scale": 120, "vertical_scale": 90})
        assert attrs["horizontal_scale"] == "120"
        assert attrs["vertical_scale"] == "90"


# ── _attrs_from_panel — language / aa_mode ───────────────────

class TestLanguageAA:
    def test_language_passes_through(self):
        attrs = _attrs_from_panel({"language": "fr"})
        assert attrs["xml_lang"] == "fr"

    def test_sharp_aa_empties(self):
        attrs = _attrs_from_panel({"anti_aliasing": "Sharp"})
        assert attrs["aa_mode"] == ""

    def test_non_default_aa(self):
        attrs = _attrs_from_panel({"anti_aliasing": "Crisp"})
        assert attrs["aa_mode"] == "Crisp"


# ── apply_character_panel_to_selection — end-to-end ──────────

class TestApplyEndToEnd:
    """Exercises the store -> attrs -> selection -> document path.
    Uses a stub Model + StateStore to keep the test hermetic."""

    def _make_model_and_store(self, text_elem):
        """Tiny Model stub with a single Text in the selection."""
        from workspace_interpreter.state_store import StateStore

        class _Sel:
            def __init__(self, path):
                self.path = path

        class _Doc:
            def __init__(self, elem):
                self._elem = elem
                self.selection = [_Sel((0, 0))]

            def get_element(self, path):
                return self._elem

            def replace_element(self, path, new_elem):
                new = _Doc(new_elem)
                new.selection = self.selection
                return new

        class _Model:
            def __init__(self, elem):
                self.document = _Doc(elem)
                self.snapshots = 0

            def snapshot(self):
                self.snapshots += 1

        return _Model(text_elem), StateStore()

    def test_font_family_written_to_selected_text(self):
        from geometry.element import Text
        t = Text(x=0, y=0, content="hello", font_family="sans-serif", font_size=12)
        model, store = self._make_model_and_store(t)
        store.init_panel("character_panel", {"font_family": "Arial", "font_size": 12})
        apply_character_panel_to_selection(store, model)
        assert model.document.get_element((0, 0)).font_family == "Arial"
        assert model.snapshots == 1

    def test_no_op_when_no_model(self):
        from workspace_interpreter.state_store import StateStore
        store = StateStore()
        store.init_panel("character_panel", {"font_family": "Arial"})
        apply_character_panel_to_selection(store, None)  # doesn't raise

    def test_no_op_when_selection_empty(self):
        from geometry.element import Text
        from workspace_interpreter.state_store import StateStore

        class _Doc:
            def __init__(self):
                self.selection = []

            def get_element(self, path):
                raise AssertionError("should not be called")

            def replace_element(self, path, new_elem):
                raise AssertionError("should not be called")

        class _Model:
            def __init__(self):
                self.document = _Doc()
                self.snapshots = 0

            def snapshot(self):
                self.snapshots += 1

        model = _Model()
        store = StateStore()
        apply_character_panel_to_selection(store, model)
        assert model.snapshots == 0

    def test_underline_flows_to_text_decoration(self):
        from geometry.element import Text
        t = Text(x=0, y=0, content="hi", font_family="serif", font_size=12)
        model, store = self._make_model_and_store(t)
        store.init_panel("character_panel", {"underline": True, "font_size": 12})
        apply_character_panel_to_selection(store, model)
        assert model.document.get_element((0, 0)).text_decoration == "underline"

    def test_subscribe_fires_on_panel_write(self):
        from geometry.element import Text
        t = Text(x=0, y=0, content="hi", font_family="serif", font_size=12)
        model, store = self._make_model_and_store(t)
        store.init_panel("character_panel", {"font_family": "sans-serif", "font_size": 12})
        subscribe(store, lambda: model)
        # The subscribe firing after every set_panel is what wires the
        # panel's write-back to the selected element.
        store.set_panel("character_panel", "font_family", "Courier New")
        assert model.document.get_element((0, 0)).font_family == "Courier New"

File: jas_ocaml/test/interpreter/character_panel_state_test.ml
(** Cross-language SEAM tests for the Character panel apply-to-selection
    pipeline. Mirrors Python's jas/panels/character_panel_state_test.py:
    every panel field -> element attribute mapping is pinned to the SAME
    input -> output values so the two apps stay equivalent.

    The pure mapping cases hit [Effects.attrs_from_character_panel]
    directly (mirrors Python's _attrs_from_panel); the end-to-end cases
    drive [Effects.apply_character_panel_to_selection] through a real
    Controller + Document (mirrors Python's TestApplyEndToEnd). *)

open Jas

(* ── helpers ─────────────────────────────────────────────── *)

let attrs (panel : (string * Yojson.Safe.t) list) =
  Effects.attrs_from_character_panel panel

let check_str name expected got = Alcotest.(check string) name expected got
let check_str_opt name expected got =
  Alcotest.(check (option string)) name expected got

(* Build a single-Text model with the given paths selected. *)
let model_with_elem elem selected =
  let layer = Element.make_layer [| elem |] in
  let selection = List.fold_left (fun acc path ->
    Document.PathMap.add path (Document.make_element_selection path) acc)
    Document.PathMap.empty selected in
  let doc = Document.make_document ~selection [| layer |] in
  Model.create ~document:doc ()

let text_font_family model path =
  match Document.get_element model#document path with
  | Element.Text t -> t.font_family
  | _ -> "<not-text>"

let text_decoration model path =
  match Document.get_element model#document path with
  | Element.Text t -> t.text_decoration
  | _ -> "<not-text>"

(* ── pure mapping: _attrs_from_panel equivalents ─────────── *)

let pure_tests = [
  (* font_size *)
  Alcotest.test_case "font_size_24" `Quick (fun () ->
    let a = attrs [("font_size", `Int 24)] in
    Alcotest.(check (option (float 1e-9))) "font_size" (Some 24.0) a.font_size);

  (* style_name -> font_weight + font_style *)
  Alcotest.test_case "style_regular" `Quick (fun () ->
    let a = attrs [("style_name", `String "Regular")] in
    check_str_opt "weight" (Some "normal") a.font_weight;
    check_str_opt "style" (Some "normal") a.font_style);
  Alcotest.test_case "style_italic" `Quick (fun () ->
    let a = attrs [("style_name", `String "Italic")] in
    check_str_opt "weight" (Some "normal") a.font_weight;
    check_str_opt "style" (Some "italic") a.font_style);
  Alcotest.test_case "style_bold" `Quick (fun () ->
    let a = attrs [("style_name", `String "Bold")] in
    check_str_opt "weight" (Some "bold") a.font_weight;
    check_str_opt "style" (Some "normal") a.font_style);
  Alcotest.test_case "style_bold_italic" `Quick (fun () ->
    let a = attrs [("style_name", `String "Bold Italic")] in
    check_str_opt "weight" (Some "bold") a.font_weight;
    check_str_opt "style" (Some "italic") a.font_style);
  Alcotest.test_case "style_unknown_leaves_none" `Quick (fun () ->
    let a = attrs [("style_name", `String "Something Weird")] in
    check_str_opt "weight" None a.font_weight;
    check_str_opt "style" None a.font_style);

  (* caps *)
  Alcotest.test_case "all_caps" `Quick (fun () ->
    let a = attrs [("all_caps", `Bool true)] in
    check_str "transform" "uppercase" a.text_transform;
    check_str "variant" "" a.font_variant);
  Alcotest.test_case "small_caps_when_all_caps_off" `Quick (fun () ->
    let a = attrs [("small_caps", `Bool true)] in
    check_str "transform" "" a.text_transform;
    check_str "variant" "small-caps" a.font_variant);
  Alcotest.test_case "all_caps_wins_over_small_caps" `Quick (fun () ->
    let a = attrs [("all_caps", `Bool true); ("small_caps", `Bool true)] in
    check_str "transform" "uppercase" a.text_transform;
    check_str "variant" "" a.font_variant);

  (* text_decoration *)
  Alcotest.test_case "decoration_none" `Quick (fun () ->
    let a = attrs [("underline", `Bool false); ("strikethrough", `Bool false)] in
    check_str "decoration" "" a.text_decoration);
  Alcotest.test_case "decoration_underline" `Quick (fun () ->
    let a = attrs [("underline", `Bool true)] in
    check_str "decoration" "underline" a.text_decoration);
  Alcotest.test_case "decoration_strikethrough" `Quick (fun () ->
    let a = attrs [("strikethrough", `Bool true)] in
    check_str "decoration" "line-through" a.text_decoration);
  Alcotest.test_case "decoration_both_alphabetical" `Quick (fun () ->
    let a = attrs [("underline", `Bool true); ("strikethrough", `Bool true)] in
    check_str "decoration" "line-through underline" a.text_decoration);

  (* leading -> line_height *)
  Alcotest.test_case "leading_auto_empty" `Quick (fun () ->
    let a = attrs [("font_size", `Int 12); ("leading", `Float 14.4)] in
    check_str "line_height" "" a.line_height);
  Alcotest.test_case "leading_explicit" `Quick (fun () ->
    let a = attrs [("font_size", `Int 12); ("leading", `Int 20)] in
    check_str "line_height" "20pt" a.line_height);

  (* tracking -> letter_spacing *)
  Alcotest.test_case "tracking_zero_empty" `Quick (fun () ->
    let a = attrs [("tracking", `Int 0)] in
    check_str "letter_spacing" "" a.letter_spacing);
  Alcotest.test_case "tracking_positive" `Quick (fun () ->
    let a = attrs [("tracking", `Int 25)] in
    check_str "letter_spacing" "0.025em" a.letter_spacing);

  (* kerning *)
  Alcotest.test_case "kerning_zero_empty" `Quick (fun () ->
    let a = attrs [("kerning", `Int 0)] in
    check_str "kerning" "" a.kerning);
  Alcotest.test_case "kerning_positive" `Quick (fun () ->
    let a = attrs [("kerning", `Int 50)] in
    check_str "kerning" "0.05em" a.kerning);
  Alcotest.test_case "kerning_numeric_string" `Quick (fun () ->
    let a = attrs [("kerning", `String "25")] in
    check_str "kerning" "0.025em" a.kerning);
  Alcotest.test_case "kerning_optical_passthrough" `Quick (fun () ->
    check_str "optical" "Optical" (attrs [("kerning", `String "Optical")]).kerning;
    check_str "metrics" "Metrics" (attrs [("kerning", `String "Metrics")]).kerning);
  Alcotest.test_case "kerning_auto_empty" `Quick (fun () ->
    check_str "auto" "" (attrs [("kerning", `String "Auto")]).kerning;
    check_str "zero" "" (attrs [("kerning", `String "0")]).kerning;
    check_str "empty" "" (attrs [("kerning", `String "")]).kerning);

  (* baseline_shift *)
  Alcotest.test_case "baseline_zero_empty" `Quick (fun () ->
    let a = attrs [("baseline_shift", `Int 0)] in
    check_str "bs" "" a.baseline_shift);
  Alcotest.test_case "baseline_numeric" `Quick (fun () ->
    let a = attrs [("baseline_shift", `Int 3)] in
    check_str "bs" "3pt" a.baseline_shift);
  Alcotest.test_case "baseline_super_wins" `Quick (fun () ->
    let a = attrs [("superscript", `Bool true); ("baseline_shift", `Int 5)] in
    check_str "bs" "super" a.baseline_shift);
  Alcotest.test_case "baseline_sub" `Quick (fun () ->
    let a = attrs [("subscript", `Bool true)] in
    check_str "bs" "sub" a.baseline_shift);

  (* rotation *)
  Alcotest.test_case "rotation_zero_empty" `Quick (fun () ->
    check_str "rot" "" (attrs [("character_rotation", `Int 0)]).rotate);
  Alcotest.test_case "rotation_numeric" `Quick (fun () ->
    check_str "rot" "15" (attrs [("character_rotation", `Int 15)]).rotate);

  (* scale *)
  Alcotest.test_case "scale_identity_empty" `Quick (fun () ->
    let a = attrs [("horizontal_scale", `Int 100); ("vertical_scale", `Int 100)] in
    check_str "h" "" a.horizontal_scale;
    check_str "v" "" a.vertical_scale);
  Alcotest.test_case "scale_non_identity" `Quick (fun () ->
    let a = attrs [("horizontal_scale", `Int 120); ("vertical_scale", `Int 90)] in
    check_str "h" "120" a.horizontal_scale;
    check_str "v" "90" a.vertical_scale);

  (* language / anti-aliasing *)
  Alcotest.test_case "language_passthrough" `Quick (fun () ->
    check_str_opt "lang" (Some "fr") (attrs [("language", `String "fr")]).xml_lang);
  Alcotest.test_case "aa_sharp_empty" `Quick (fun () ->
    check_str "aa" "" (attrs [("anti_aliasing", `String "Sharp")]).aa_mode);
  Alcotest.test_case "aa_non_default" `Quick (fun () ->
    check_str "aa" "Crisp" (attrs [("anti_aliasing", `String "Crisp")]).aa_mode);
]

(* ── end-to-end apply-to-selection ───────────────────────── *)

let e2e_tests = [
  Alcotest.test_case "font_family_written_to_selected_text" `Quick (fun () ->
    let t = Element.make_text ~font_family:"sans-serif" ~font_size:12.0
              0.0 0.0 "hello" in
    let model = model_with_elem t [[0; 0]] in
    let store = State_store.create () in
    State_store.init_panel store "character_panel_content"
      [("font_family", `String "Arial"); ("font_size", `Int 12)];
    let ctrl = new Controller.controller ~model () in
    Effects.apply_character_panel_to_selection store ctrl;
    check_str "font_family" "Arial" (text_font_family model [0; 0]));

  Alcotest.test_case "underline_flows_to_text_decoration" `Quick (fun () ->
    let t = Element.make_text ~font_family:"serif" ~font_size:12.0
              0.0 0.0 "hi" in
    let model = model_with_elem t [[0; 0]] in
    let store = State_store.create () in
    State_store.init_panel store "character_panel_content"
      [("font_family", `String "serif"); ("underline", `Bool true);
       ("font_size", `Int 12)];
    let ctrl = new Controller.controller ~model () in
    Effects.apply_character_panel_to_selection store ctrl;
    check_str "decoration" "underline" (text_decoration model [0; 0]));

  Alcotest.test_case "no_op_when_selection_empty" `Quick (fun () ->
    let t = Element.make_text ~font_family:"serif" ~font_size:12.0
              0.0 0.0 "hi" in
    let model = model_with_elem t [] in
    let store = State_store.create () in
    State_store.init_panel store "character_panel_content"
      [("font_family", `String "Arial"); ("font_size", `Int 12)];
    let ctrl = new Controller.controller ~model () in
    Effects.apply_character_panel_to_selection store ctrl;
    (* Element untouched: no selected text to write to. *)
    check_str "font_family" "serif" (text_font_family model [0; 0]));
]

let () =
  Alcotest.run "character_panel_state" [
    "pure_mapping", pure_tests;
    "end_to_end", e2e_tests;
  ]
